### PR TITLE
Optimize multiplication and make it constant time

### PR DESCRIPTION
**Pull request by wbl**

### CODE REVIEW

File: ecc/bls12381/g1.go [hunk 1]
```diff
@@ -174,10 +182,40 @@ func (g *G1) Add(P, Q *G1) {
 // ScalarMult calculates g = kP.
 func (g *G1) ScalarMult(k *Scalar, P *G1) { g.scalarMult(k.Bytes(), P) }
 
+// ScalarMultVarTime calculates g=kP, leaking k, for short k
+func (g *G1) ScalarMultVarTimeShort(k *Scalar, P *G1) { g.scalarMultVarTimeShort(k.Bytes(), P) }
```
Comment (armfazh) [inline ecc/bls12381/g1.go:186]: `ScalarMultVarTimeShort`--> `ScalarMultVarTime`

Not sure if we want to expose this function

File: ecc/bls12381/g1.go [hunk 2]
```diff
@@ -174,10 +182,40 @@ func (g *G1) Add(P, Q *G1) {
...
+		idx := 0xf & (k[i/8] >> uint(i%8))
+		for j := 0; j < 16; j++ {
+			T.cmov(&mults[j], subtle.ConstantTimeByteEq(idx, uint8(j)))
+		}
+		Q.Add(&Q, T)
+	}
+	g.Set(&Q)
+}
+
+// Because the arguments are so short, nothing helps
```
Comment (armfazh) [inline ecc/bls12381/g1.go:214]: is this part of the documentation of this function? if so, better to be more explicit

File: ecc/bls12381/g1.go (hunk 1 above)
Comment (wbl) [reply ecc/bls12381/g1.go:186]: let's keep it internal. I added short to emphasize that the tradeoffs are different.

File: ecc/bls12381/g1.go [hunk 3]
```diff
@@ -177,7 +185,35 @@ func (g *G1) ScalarMult(k *Scalar, P *G1) { g.scalarMult(k.Bytes(), P) }
...
+			T.cmov(&mults[j], subtle.ConstantTimeByteEq(idx, uint8(j)))
+		}
+		Q.Add(&Q, T)
+	}
+	g.Set(&Q)
+}
+
+// scalarMultShort multiplies by short, constant scalars.
+func (g *G1) scalarMultShort(k []byte, P *G1) {
+	// Since the scalar is short and low Hamming weight not much helps.
```
Comment (bwesterb) [inline ecc/bls12381/g1.go:213]: Addition chains might help — in this case I found one with one double less, which is not worth the trouble.

**Review by bwesterb — APPROVED** [review]

**Review by armfazh — APPROVED** [review]
Agree with Bas' comments.

**Pull request merged**